File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int first = binarySearch(nums, target, true);
        if (first == -1) return {-1, -1}; // If target is not found, return [-1, -1]
        int last = binarySearch(nums, target, false);
        return {first, last};
    }
    int binarySearch(vector<int>& nums, int target, bool findFirst) {
        int low = 0, high = nums.size() - 1, result = -1;
        while (low <= high) {
            int mid = low + (high - low) / 2; // Avoid potential overflow
            if (nums[mid] == target) {
                result = mid;
                if (findFirst)
                    high = mid - 1; // Search left half for first occurrence
                else
                    low = mid + 1; // Search right half for last occurrence
            } else if (nums[mid] < target) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return result;
    }
};

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int LeftMost(int arr[],int target){
        int l=0;
        int r=arr.length-1;
        while(l<=r){
            int m=l+(r-l)/2;
            if(arr[m]>=target){
                r=m-1;
            }
            else{
                l=m+1;
            }
        }
        if(l>=arr.length){
            return -1;
        }
        if(arr[l]!=target){
            return -1;
        }
        return l;
    }
   public int RightMost(int arr[],int target){
    int l=0;
    int r=arr.length-1;
    while(l<=r){
        int m=l+(r-l)/2;
        if(arr[m]>target){
            r=m-1;
        }else{
            l=m+1;
        }
    }
    if(r<0){
        return -1;
    }
    if(arr[r]!=target){
        return -1;

    }
    return r;
   }

    public int[] searchRange(int[] nums, int target) {
        int lm=LeftMost(nums,target);
        int rm=RightMost(nums,target);
        int ans[]={lm,rm};
        return ans;
    }
}